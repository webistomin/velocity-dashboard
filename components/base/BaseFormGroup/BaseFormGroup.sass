+b('base-form-group')
  +e('label')
    display: block

  +e('heading')
    display: flex
    justify-content: space-between
    align-items: baseline
    flex-wrap: wrap
    margin-bottom: rem(8px)

  +e('input')
    box-sizing: border-box
    width: 100%
    padding: rem(10px) rem(16px)
    font: var(--text-input)
    color: rgb(var(--color-primary))
    background-color: rgba(var(--color-lavender), 0.2)
    border: rem(1px) solid rgb(var(--color-lavender))
    border-radius: 5px
    transition: border-color 0.2s ease-out, color 0.2s ease-out
    outline: none

    &:focus
      border-color: rgb(var(--color-primary))
      caret-color: rgb(var(--color-primary))

      &::placeholder
        color: rgb(var(--color-primary))

    &:hover
      border-color: rgb(var(--color-primary))

    +m('invalid')
      border-color: rgb(var(--color-red))
      color: rgb(var(--color-red))

      &::placeholder
        color: inherit

  +e('errors')
    display: flex
    flex-direction: column
    padding-top: rem(4px)

  +e('error')
    font: var(--text-caption)
    color: rgb(var(--color-red))

    &:not(:last-of-type)
      margin-bottom: rem(4px)

body.using-keyboard .base-form-group__input:focus
  outline: rem(1px) solid rgb(var(--color-primary))
