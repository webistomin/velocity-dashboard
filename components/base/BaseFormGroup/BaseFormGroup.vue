<template lang="pug">
  .base-form-group
    .base-form-group__heading
      label(
        v-if='label'
        :for='id'
      ).base-form-group__label.caption {{ label }}
      slot(name='heading')
    .base-form-group__content
      input(
        :type='type'
        :id='id'
        :aria-label='label'
        :name='name'
        @input='onInput($event)'
        @blur='onBlur'
        :value='value'
        :aria-describedby='`errors-${id}`'
        :autocomplete='autocomplete'
        :placeholder='placeholder'
        :class='{"base-form-group__input_invalid" : isInputInvalid}'
      ).base-form-group__input
      ul.base-form-group__errors.list(
        :id='`errors-${id}`'
        v-if='isInputInvalid'
        )
        li(
          v-for='(error, index) in activeErrorMessages'
          :key='`${error}-${index}`'
        ).base-form-group__error.list-item {{ error }}

</template>

<script lang="ts">
  import Vue from 'vue';
  import Component, { mixins } from 'vue-class-component';
  // @ts-ignore
  import { singleErrorExtractorMixin } from 'vuelidate-error-extractor';

  const BaseFormGroupProps = Vue.extend({
    props: {
      type: {
        type: String,
        default: 'text',
      },
      label: {
        type: String,
        default: '',
      },
      id: {
        type: String,
        default: '',
      },
      value: {
        type: [String, Number],
        default: '',
        required: true,
      },
      name: {
        type: String,
        default: '',
      },
      placeholder: {
        type: String,
        default: '',
      },
      autocomplete: {
        type: String,
        default: '',
      },
    },
  });

  @Component({
    name: 'BaseFormGroup',
  })
  export default class BaseFormGroup extends mixins(BaseFormGroupProps, singleErrorExtractorMixin) {
    public activeErrorMessages!: string[];

    public get isInputInvalid(): boolean {
      return Boolean(this.activeErrorMessages && this.activeErrorMessages.length);
    }

    public onInput(event: KeyboardEvent): void {
      const target = event.target as HTMLInputElement;
      this.$emit('input', target.value);
    }

    public onBlur(): void {
      this.$emit('blur');
    }
  }
</script>

<style lang="sass">
  @import "BaseFormGroup"
</style>
