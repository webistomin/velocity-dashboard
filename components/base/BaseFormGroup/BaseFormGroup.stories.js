import { text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { email, required } from 'vuelidate/lib/validators';
import BaseFormGroup from './BaseFormGroup';

export default {
  title: 'BaseFormGroup',
  parameters: {
    options: {
      showPanel: true,
      selectedPanel: 'storybook/knobs/panel',
      previewTabs: {
        actions: {
          hidden: true,
        },
      },
    },
  },
};

export const Default = () => ({
  components: { BaseFormGroup },
  props: {
    type: {
      type: String,
      default: text('type', 'text'),
    },
    label: {
      type: String,
      default: text('label', 'label'),
    },
    id: {
      type: String,
      default: text('id', 'base-form-group'),
    },
    name: {
      type: String,
      default: text('name', ''),
    },
    placeholder: {
      type: String,
      default: text('placeholder', 'placeholder'),
    },
    autocomplete: {
      type: String,
      default: text('autocomplete', ''),
    },
  },
  data() {
    return {
      signInForm: {
        email: '',
        password: '',
      },
    };
  },
  validations: {
    signInForm: {
      email: {
        required,
        email,
      },
      password: {
        required,
      },
    },
  },
  methods: {
    onInput: action('emit input'),
    onBlur: action('emit blur'),
  },
  template: `<base-form-group :type='this.type'
                              :label='this.label'
                              :id='this.id'
                              :placeholder='this.placeholder'
                              :validator='$v.signInForm.email'
                              @input='signInForm.email = $event, onInput()'
                              :value='signInForm.email'
                              @blur='$v.signInForm.email.$touch(), onBlur()'/>`,
});
