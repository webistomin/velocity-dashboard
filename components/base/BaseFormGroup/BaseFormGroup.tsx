import { Component, Prop, Emit } from 'nuxt-property-decorator';
import { VueComponent } from 'types/vue-components';
import { VNode } from 'vue';
// @ts-ignore
import { singleErrorExtractorMixin } from 'vuelidate-error-extractor';

import { TheMask } from 'vue-the-mask';

import './BaseFormGroup.sass';

export type BaseFormGroupTypes =
  | 'email'
  | 'date'
  | 'datetime-local'
  | 'file'
  | 'month'
  | 'number'
  | 'password'
  | 'search'
  | 'tel'
  | 'text'
  | 'time'
  | 'url'
  | 'week';

export interface IBaseFormGroupProps {
  type?: BaseFormGroupTypes;
  label?: string;
  id?: string;
  value: string | number;
  name?: string;
  placeholder?: string;
  autocomplete?: string;
  onInput: () => string | number;
  onBlur?: () => void;
  mask?: string;
}

@Component({
  name: 'BaseFormGroup',
  mixins: [singleErrorExtractorMixin],
})
export default class BaseFormGroup extends VueComponent<IBaseFormGroupProps> {
  public activeErrorMessages!: string[];

  @Prop({ default: 'text' })
  private readonly type!: IBaseFormGroupProps['type'];

  @Prop()
  private readonly label!: IBaseFormGroupProps['label'];

  @Prop()
  private readonly id!: IBaseFormGroupProps['id'];

  @Prop()
  private readonly value!: IBaseFormGroupProps['value'];

  @Prop()
  private readonly name!: IBaseFormGroupProps['name'];

  @Prop()
  private readonly placeholder!: IBaseFormGroupProps['placeholder'];

  @Prop()
  private readonly autocomplete!: IBaseFormGroupProps['autocomplete'];

  @Prop()
  private readonly mask!: IBaseFormGroupProps['mask'];

  public get isInputInvalid(): boolean {
    return Boolean(this.activeErrorMessages && this.activeErrorMessages.length);
  }

  @Emit('input')
  public onInput(event: KeyboardEvent): string | number {
    const target = event.target as HTMLInputElement;
    return target.value;
  }

  @Emit('input')
  public onMaskedInput(value: string | number): string | number {
    return value;
  }

  @Emit('blur')
  public onBlur(): void {}

  render(): VNode {
    return (
      <div class='base-form-group'>
        <div class='base-form-group__heading'>
          {this.label ? (
            <label for={this.id} class='base-form-group__label caption label'>
              {this.label}
            </label>
          ) : null}
          {this.$slots.heading}
        </div>
        <div class='base-form-group__content'>
          {this.mask ? (
            <TheMask
              type={this.type}
              id={this.id}
              aria-label={this.label}
              name={this.name}
              onInput={this.onMaskedInput}
              nativeOnBlur={this.onBlur}
              value={this.value}
              aria-describedby={`errors-${this.id}`}
              autoComplete={this.autocomplete}
              placeholder={this.placeholder}
              class={`${this.isInputInvalid ? 'base-form-group__input_invalid' : ''} base-form-group__input`}
              mask={this.mask}
              masked
            />
          ) : (
            <input
              type={this.type}
              id={this.id}
              aria-label={this.label}
              name={this.name}
              onInput={this.onInput}
              onBlur={this.onBlur}
              value={this.value}
              aria-describedby={`errors-${this.id}`}
              autoComplete={this.autocomplete}
              placeholder={this.placeholder}
              class={`${this.isInputInvalid ? 'base-form-group__input_invalid' : ''} base-form-group__input`}
            />
          )}
          {this.isInputInvalid ? (
            <ul class='base-form-group__errors list' id={`errors-${this.id}`}>
              {this.activeErrorMessages.slice(0, 1).map((error: string, index: number) => {
                return (
                  <li class='base-form-group__error list-item' key={`${error}-${index}`}>
                    {error}
                  </li>
                );
              })}
            </ul>
          ) : null}
        </div>
      </div>
    );
  }
}
