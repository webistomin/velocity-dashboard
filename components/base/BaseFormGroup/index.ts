import BaseFormGroup from './BaseFormGroup';

export default BaseFormGroup;
