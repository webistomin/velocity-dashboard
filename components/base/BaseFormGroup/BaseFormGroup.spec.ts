import { mount } from '@vue/test-utils';
import BaseFormGroup from './BaseFormGroup';

describe('BaseFormGroup', () => {
  const inputSelector = "[data-jest='base-form-group__input']";
  const maskSelector = "[data-jest='base-form-group__mask']";
  const textareaSelector = "[data-jest='base-form-group__textarea']";
  const labelSelector = "[data-jest='base-form-group__label']";
  const errorSelector = "[data-jest='base-form-group__error']";

  const factory = (propsData = {}, options = {}) => {
    return mount(BaseFormGroup, {
      mocks: {
        $vuelidateErrorExtractor: {
          validationKeys: {},
        },
      },
      propsData: {
        value: '',
        validator: {
          required: true,
          email: true,
          $model: '',
          $invalid: true,
          $dirty: true,
          $anyDirty: true,
          $error: true,
          $anyError: true,
          $pending: false,
          $params: {
            required: {
              type: 'required',
            },
            email: {
              type: 'email',
            },
          },
        },
        ...propsData,
      },
      ...options,
      stubs: ['TheMask'],
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Emit "blur" event', () => {
    const wrapper = factory();
    const input = wrapper.find(inputSelector);
    input.trigger('blur');
    expect(wrapper.emitted().blur);
  });

  it('Emit "input" event', () => {
    const wrapper = factory();
    const input = wrapper.find(inputSelector);
    input.trigger('input');
    expect(wrapper.emitted().input);
  });

  it('emit "input" event using mask input', () => {
    const wrapper = factory({
      mask: '##/##/##',
    });
    const mask = wrapper.find(maskSelector);
    mask.trigger('input');
    expect(wrapper.emitted().input);
  });

  it('Set textarea height on input event', () => {
    const wrapper = factory({
      isTextarea: true,
    });
    const textareaEl = wrapper.vm.$refs.baseTextarea as HTMLTextAreaElement;
    const initialTextareaHeight = textareaEl.style.height;

    const textarea = wrapper.find(textareaSelector);
    textarea.setValue(
      `Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
      when an unknown printer took a galley of type and scrambled it to make a type
      specimen book. It has survived not only five centuries, but also the leap into
      electronic typesetting, remaining essentially unchanged`
    );
    textarea.trigger('input');

    const newTextareaHeight = textareaEl.style.height;
    expect(initialTextareaHeight).not.toEqual(newTextareaHeight);
  });

  it('Set label if label prop is provided', () => {
    const expectedLabel = 'label';
    const wrapper = factory({
      label: expectedLabel,
    });
    const label = wrapper.find(labelSelector);
    expect(label.exists()).toBe(true);
    expect(label.text()).toEqual(expectedLabel);
  });

  it('Hide label if label prop is not provided', () => {
    const wrapper = factory();
    const label = wrapper.find(labelSelector);
    expect(label.exists()).toBe(false);
  });

  it('Add invalid class to input if it is invalid', () => {
    const expectedResult = ['base-form-group__input_invalid'];
    const wrapper = factory(
      {},
      {
        computed: {
          isInputInvalid: () => true,
        },
      }
    );
    const input = wrapper.find(inputSelector);
    const classes = input.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });

  it('Add invalid class to textarea if it is invalid', () => {
    const expectedResult = ['base-form-group__input_invalid'];
    const wrapper = factory(
      {
        isTextarea: true,
      },
      {
        computed: {
          isInputInvalid: () => true,
        },
      }
    );
    const input = wrapper.find(textareaSelector);
    const classes = input.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });

  it('Add invalid class to mask if it is invalid', () => {
    const expectedResult = ['base-form-group__input_invalid'];
    const wrapper = factory(
      {
        mask: '##/##/##',
      },
      {
        computed: {
          isInputInvalid: () => true,
        },
      }
    );
    const mask = wrapper.find(maskSelector);
    const classes = mask.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });

  it('Show error if array is not empty', () => {
    const wrapper = factory(
      {},
      {
        computed: {
          isInputInvalid: () => true,
          activeErrorMessages: {
            get() {
              return ['Error'];
            },
            set(value: string) {
              console.log(value);
            },
          },
        },
      }
    );
    const error = wrapper.find(errorSelector);
    expect(error.exists()).toBe(true);
  });
});
