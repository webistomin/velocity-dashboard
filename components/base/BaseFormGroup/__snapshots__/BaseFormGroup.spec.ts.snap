// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseFormGroup Matches snapshot 1`] = `
"<div class=\\"base-form-group\\">
  <div class=\\"base-form-group__heading\\"></div>
  <div class=\\"base-form-group__content\\"><input data-jest=\\"base-form-group__input\\" type=\\"text\\" aria-describedby=\\"errors-undefined\\" class=\\" base-form-group__input\\"></div>
</div>"
`;
